File: Python/string-1/combo_string.py
'''
CodingBat Python Activity "combo_string" from string-1.
codingbat.com
'''

failures = 0

def combo_string(a, b):
    '''

Given 2 strings, a and b, return a string of the form short+long+short, with the shorter string on the outside and the longer string on the inside. The strings will not be the same length, but they may be empty (length 0).


combo_string('Hello', 'hi') → 'hiHellohi'
combo_string('hi', 'Hello') → 'hiHellohi'
combo_string('aaa', 'b') → 'baaab'
'''
    pass





def catchFailures(testCase, result):
    global failures
    
    try:
        assert eval(testCase) == result, "Test Case Failed: "+testCase+ " → "+str(result)
    except Exception as e:
        failures += 1
        print(e)
            


def test():
    global failures
    
    catchFailures("combo_string('Hello', 'hi')", 'hiHellohi')
    catchFailures("combo_string('hi', 'Hello')", 'hiHellohi')
    catchFailures("combo_string('aaa', 'b')", 'baaab')
    catchFailures("combo_string('b', 'aaa')", 'baaab')
    catchFailures("combo_string('aaa', '')", 'aaa')
    catchFailures("combo_string('', 'bb')", 'bb')
    catchFailures("combo_string('aaa', '1234')", 'aaa1234aaa')
    catchFailures("combo_string('aaa', 'bb')", 'bbaaabb')
    catchFailures("combo_string('a', 'bb')", 'abba')
    catchFailures("combo_string('bb', 'a')", 'abba')
    catchFailures("combo_string('xyz', 'ab')", 'abxyzab')

        
    if failures > 0:
        print("\r\n "+ str(failures) + " failed test case(s)   :(")
    else:
        print("Good job -- problem solved!")


test()

File: Java/String1/ComboString.java
/*
 * CodingBat Java Activity "comboString" from String-1 section.
 * codingbat.com
 */

public class ComboString 
{
	private static int failures = 0;
	private static final String METHOD_NAME = "comboString";
	
	/*
	 * INSTRUCTIONS
	 * 
	 * Given 2 strings, a and b, return a string of the form short+long+short, 
	 * with the shorter string on the outside and the longer string on the 
	 * inside. The strings will not be the same length, but they may be empty 
	 * (length 0).
	 * 
	 * comboString("Hello", "hi") -> "hiHellohi"
	 * comboString("hi", "Hello") -> "hiHellohi"
	 * comboString("aaa", "b") -> "baaab"
	 */
	public static String comboString(String a, String b)
	{
		//put your code ONLY here!
	}
	
	public static void main(String[] args)
	{
		testCase("\"Hello\"", "\"hi\"", "\"hiHellohi\"");
		testCase("\"hi\"", "\"Hello\"", "\"hiHellohi\"");
		testCase("\"aaa\"", "\"b\"", "\"baaab\"");
		testCase("\"b\"", "\"aaa\"", "\"baaab\"");
		testCase("\"aaa\"", "\"\"", "\"aaa\"");
		testCase("\"\"", "\"bb\"", "\"bb\"");
		testCase("\"aaa\"", "\"1234\"", "\"aaa1234aaa\"");
		testCase("\"aaa\"", "\"bb\"", "\"bbaaabb\"");
		testCase("\"a\"", "\"bb\"", "\"abba\"");
		testCase("\"bb\"", "\"a\"", "\"abba\"");
		testCase("\"xyz\"", "\"ab\"", "\"abxyzab\"");
		
		System.out.println();
		System.out.println(failures + " failed test cases.");
	}

	private static void testCase(String a, String b, String result)
	{
		String r;
		try
		{
			r = ""+ "\"" + comboString(a.substring(1,a.length()-1), b.substring(1,b.length()-1)) + "\"";
			if (r.equals(""+result))
			{
				System.out.print("PASS: ");
			}
			else
			{
				System.out.println("FAIL: ");
				failures++;
			}
		}
		catch (Exception e)
		{
			r = e.getMessage();
			System.out.print("FAIL: ");
			failures++;
		}
		System.out.println(METHOD_NAME+"("+a+", "+b+") -> "+r + ", EXPECTED: "+result);
	}

}

